HERD Updates

mavaylon1 opened the pull request:
## Motivation

What was the reasoning behind this change? Please explain the changes briefly.

- [x] HERD should resolve the reuse of keys when it is being associated with the same object. This will allow us to bulk add but also gives the user to still manually choose a key that exists if they want to use that key later on.
- [x] HERD should do checks before populating the tables. Currently, the checks are spread out (some we may not be able to move so TBD). This leads to some tables being populated and some are not when an error is thrown. If the checks can't be moved as easily as I think, then we would need to have something that removes what was added. I prefer the former to the latter.
- [x] HERD should rename add_ref_termset to maybe "add_ref_container" since it is only called in the I/O for write.
- [x] HERD should have a add_ref_termset that actually uses termsets instead of entity id or uri. We had this prior but this changed when we introduced the wrapper. This would bring the essence of that back.
- [x] HERD should resolve existing entities with a warning and use the uri of current entity and not the parameter.
- [x] Add tests for coverage
- [x] Update tutorial to account for add_ref_termset

## How to test the behavior?
```
Show how to reproduce the new behavior (can be a bug fix or a new feature)
```
Unit Tests

## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [x] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR clearly describes the problem and the solution?
- [x] Is your contribution compliant with our coding style? This can be checked running `ruff` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/hdmf-dev/hdmf/pulls) for the same change?
- [x] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

mavaylon1 committed:
HERD Updates

mavaylon1 committed:
rework in progress

mavaylon1 committed:
updated tests/go through for final checks

mavaylon1 committed:
Merge branch 'dev' into herd_update

mavaylon1 committed:
add_ref_termset and tests

mavaylon1 committed:
comments

mavaylon1 committed:
comments

mavaylon1 committed:
gallery

mavaylon1 committed:
checks

mavaylon1 committed:
Merge branch 'dev' into herd_update

mavaylon1 committed:
path

mavaylon1 committed:
path

mavaylon1 committed:
path

mavaylon1 committed:
cov

mavaylon1 committed:
cov

mavaylon1 committed:
doc

mavaylon1 committed:
notes'

mavaylon1 committed:
Update CHANGELOG.md

mavaylon1 commented:
Fix #961

mavaylon1 committed:
resuse uri

mavaylon1 committed:
typo

mavaylon1 committed:
edits

mavaylon1 commented:
This PR has grown to cover multiple items.

- add_ref_termset
- resolving keys within the same object
- resolving entities to ignore new a uri

I remembered a question about `get_key` and I want to get better at limiting the scope of PRs.  I will continue off of the original PR: https://github.com/hdmf-dev/hdmf/pull/964 
The idea is that `get_key` should: 

- If no key is found return None instead of raising a ValueError
- If multiple matching keys are found, then return a list of Key objects rather than raising a ValueError
I don't think changing to None is more informative. I like the fact the error tells you why. As for the second one, I think that's fine.

mavaylon1 committed:
Merge branch 'dev' into herd_update

mavaylon1 committed:
Update CHANGELOG.md

mavaylon1 committed:
Merge branch 'dev' into herd_update

mavaylon1 committed:
Merge branch 'dev' into herd_update

mavaylon1 committed:
Merge branch 'dev' into herd_update

rly commented on docs/gallery/plot_external_resources.py:
```diff
@@ -306,6 +307,52 @@ def __init__(self, **kwargs):
+# ------------------------------------------------------
+# :py:class:`~hdmf.common.resources.HERD` has multiple ways for users to add
+# external references. The :py:func:`~hdmf.common.resources.HERD.add_ref_termset`
+# method allows users to not only validate terms, i.e., keys, but also adds the
+# ability to iteratively add references for entire datasets, lists, arrays, etc.
```
```suggestion
# method allows users to not only validate terms, i.e., keys, but also 
# add references for entire datasets, lists, arrays, etc.
```
Fix grammar and improve clarity

rly commented on docs/gallery/plot_external_resources.py:
```diff
@@ -306,6 +307,52 @@ def __init__(self, **kwargs):
+# method allows users to not only validate terms, i.e., keys, but also adds the
+# ability to iteratively add references for entire datasets, lists, arrays, etc.
+
+# The :py:func:`~hdmf.common.resources.HERD.add_ref_container` method is directly
+# used for populating :py:class:`~hdmf.common.resources.HERD` when writing a file.
+# :py:func:`~hdmf.common.resources.HERD.add_ref_termset` can be used for new files;
+# however, it is also the best practice when adding references for existing files.
```
I'm a bit confused here. When should a user use `add_ref_container`? Or should that be used only internally?

oruebel replied:
As far as I can tell, I think `add_ref_container` in the code is right now used only internally. I don't think `add_ref_container` applies to existing files, as that would require wrapping fields in Containers that have been read from disk with TermSetWrapper, which I don't think is supported right now. For existing files, one would use `add_ref_termset`.

oruebel replied:
I'm not sure if/when a user would need to call ``add_ref_container`` directly.

rly commented on docs/gallery/plot_external_resources.py:
```diff
@@ -306,6 +307,52 @@ def __init__(self, **kwargs):
+
+# :py:func:`~hdmf.common.resources.HERD.add_ref_termset` has many optional fields,
+# giving the user a range of control when adding references. Let's see an example.
+er = HERD()
```
```suggestion
herd = HERD()
```
Can we name the variable `er` to `herd` throughout these examples? I don't think `er` would mean anything to a new user, now that we have renamed everything to HERD.

rly commented on docs/gallery/plot_external_resources.py:
```diff
@@ -306,6 +307,52 @@ def __init__(self, **kwargs):
+
+# When populating :py:class:`~hdmf.common.resources.HERD`, users may have some terms
+# that are not in the :py:class:`~hdmf.term_set.TermSet`. As a result,
+# :py:func:`~hdmf.common.resources.HERD.add_ref_termset` will return an all the missing
+# terms in a dictionary. It is up to the user to either add these terms to the
+# :py:class:`~hdmf.term_set.TermSet` or remove them from the dataset.
```
I thought `add_ref_termset` would raise a validation error if a user tries to add terms not in the term set. Is that not correct?

mavaylon1 replied:
No. `add_ref_termset` returns all the missing terms. This way when you bulk add, you see all the bad data vs having to correct one bad data point at a time.

You are thinking of the error being raised within the TermSet class itself.

rly commented on docs/gallery/plot_external_resources.py:
```diff
@@ -306,6 +307,52 @@ def __init__(self, **kwargs):
+# giving the user a range of control when adding references. Let's see an example.
+er = HERD()
+terms = TermSet(term_schema_path='example_term_set.yaml')
+file = HERDManagerContainer(name='file')
```
Why do we have to create a new file here? If we create a new file, should we also create a new container? Right now the species Data container isn't even in this file. Can we add a sanity check in `add_ref_termset` that if `file` and `container` are both passed that `container` is actually part of the `file`? This could be an easy way for users to accidentally create nonsense references. I think this example would create a nonsense reference.

mavaylon1 replied:
good idea

rly commented on src/hdmf/common/resources.py:
```diff
@@ -358,16 +361,17 @@ def _check_object_field(self, **kwargs):
         file_object_id = file.object_id
         files_idx = self.files.which(file_object_id=file_object_id)
 
         if len(files_idx) > 1:
```
```suggestion
        if len(files_idx) > 1:  # pragma: no cover
```

rly commented on src/hdmf/common/resources.py:
```diff
@@ -378,10 +382,16 @@ def _check_object_field(self, **kwargs):
+                    'field': field}
         elif len(objecttable_idx) == 0 and not create:
             raise ValueError("Object not in Object Table.")
         else:
```
```suggestion
        else:  # pragma: no cover
```

rly commented on src/hdmf/common/resources.py:
```diff
@@ -466,63 +476,69 @@ def add_ref_term_set(self, **kwargs):
-        If container, relative_path, and field are provided, the Key that corresponds to the given name of the key
-        for the given container, relative_path, and field is returned.
+        This method allows users to take advantage of using the TermSet class to provide the entity information
+        for add_ref, while also validating the data. This method supports adding a single key and an entire dataset
```
```suggestion
        for add_ref, while also validating the data. This method supports adding a single key or an entire dataset
```

rly commented on src/hdmf/common/resources.py:
```diff
@@ -466,63 +476,69 @@ def add_ref_term_set(self, **kwargs):
+        for add_ref, while also validating the data. This method supports adding a single key and an entire dataset
+        to the HERD tables. For both cases, the term, i.e., key, will be validated against the permissible values
+        in the TermSet. If valid, it will proceed to call add_ref. Otherwise, the method will return a dict of
+        missing terms.
```
```suggestion
        missing terms (terms not found in the TermSet).
```

oruebel commented on CHANGELOG.md:
```diff
@@ -2,6 +2,9 @@
 ## HDMF 3.11.1 (Upcoming)
 
+### Enhancements
+- Added `add_ref_termset`, updated helper methods for `HERD`, and revised `add_ref` to support validations prior to populating the tables. @mavaylon1 [#968](https://github.com/hdmf-dev/hdmf/pull/968)
```
I believe `add_ref_container` is also new and should be mentioned here.

rly commented on src/hdmf/common/resources.py:
```diff
@@ -466,63 +476,69 @@ def add_ref_term_set(self, **kwargs):
+            elif isinstance(data_object, (list, tuple, np.ndarray)):
+                data = data_object
+        missing_terms = []
+        for term in data:
```
There is a possibility that `data` is not set in the code above. e.g., if `data` is a DataChunkIterator or an h5py.Dataset. Can you add an else statement below 522 that raises an error that says something like "data_object is not a Data, DataIO, list, tuple, or numpy ndarray and cannot be used in add_ref_termset"?

rly commented on src/hdmf/common/resources.py:
```diff
@@ -466,63 +476,69 @@ def add_ref_term_set(self, **kwargs):
+                         entity_id=entity_id,
+                         entity_uri=entity_uri)
+        if len(missing_terms)>0:
+            return {"missing_terms": missing_terms}
```
I forget - why do we return a dict here with a single key-value pair instead of just the list of missing terms?

mavaylon1 replied:
My personal preference in that it gives information that it is indeed missing terms

rly commented on src/hdmf/common/resources.py:
```diff
@@ -605,69 +667,141 @@ def add_ref(self, **kwargs):
+                        obj_key_exists = True
+                        # this means there is already a object-key relationship recorded
+                if not obj_key_exists:
+                    # this means that though the key is there, there is not object-key relationship
```
```suggestion
                    # this means that though the key is there, there is no object-key relationship
```

rly commented on src/hdmf/common/resources.py:
```diff
@@ -605,69 +667,141 @@ def add_ref(self, **kwargs):
-                    # this means that though the key is there, there is not key-entity relationship
-                    # a.k.a add it now
-                    self._add_entity_key(entity, key)
+                    # this means that though the key is there, there is not entity-key relationship
```
```suggestion
                    # this means that though the key is there, there is no entity-key relationship
```

rly commented:
Thanks for all this work and sorry for my delayed review. Overall it looks good. I made some comments above.

I also think we should change the examples/tutorial to be more realistic. Right now, you create a `HERDManagerContainer` object to represent a file, but in the later examples, the file object and the file of the data object do not match each other, and I think that is confusing.

```python
# Class to represent a file
class HERDManagerContainer(Container, HERDManager):
    def __init__(self, **kwargs):
        kwargs['name'] = 'HERDManagerContainer'
        super().__init__(**kwargs)
```

Instead, I suggest using a `SimpleMultiContainer` as the base for the file, and in each example, create a new file, create a new data object, and add the data object to the file before using it:
```python
# Class to represent a file
from hdmf.common import SimpleMultiContainer
class HdmfFile(SimpleMultiContainer, HERDManager):
    def __init__(self, **kwargs):
        kwargs['name'] = 'SimpleMultiContainer'
        super().__init__(**kwargs)

...

file = HdmfFile()
species = DynamicTable(name='species', description='My species', columns=[col1])
file.add_container(species)
```

Then in the examples, the `file` argument does not need to be passed because it can be retrieved from the `data` object.
```python
er.add_ref_termset(
	container=species,
	attribute='Species_Data',
	key='Ursus arctos horribilis',
	termset=terms
)
```
This is a more realistic use case, in my opinion. This structure also has the advantage of making the code examples being more standalone / not dependent on earlier objects, except for the definition of `HdmfFile`.

All of this would help make this tutorial more analogous to its use in PyNWB, so that we could copy and paste relevant parts of this tutorial to create a PyNWB tutorial with minimal modifications, e.g., `HdmfFile` -> `NWBFile`. `add_container` -> `add_acquisition`. Perhaps all of this should be a separate PR though. Up to you. Happy to help with these suggested changes.

mavaylon1 committed:
Update docs/gallery/plot_external_resources.py

Co-authored-by: Ryan Ly <rly@lbl.gov>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

mavaylon1 commented:
> Thanks for all this work and sorry for my delayed review. Overall it looks good. I made some comments above.
> 
> I also think we should change the examples/tutorial to be more realistic. Right now, you create a `HERDManagerContainer` object to represent a file, but in the later examples, the file object and the file of the data object do not match each other, and I think that is confusing.
> 
> ```python
> # Class to represent a file
> class HERDManagerContainer(Container, HERDManager):
>     def __init__(self, **kwargs):
>         kwargs['name'] = 'HERDManagerContainer'
>         super().__init__(**kwargs)
> ```
> 
> Instead, I suggest using a `SimpleMultiContainer` as the base for the file, and in each example, create a new file, create a new data object, and add the data object to the file before using it:
> 
> ```python
> # Class to represent a file
> from hdmf.common import SimpleMultiContainer
> class HdmfFile(SimpleMultiContainer, HERDManager):
>     def __init__(self, **kwargs):
>         kwargs['name'] = 'SimpleMultiContainer'
>         super().__init__(**kwargs)
> 
> ...
> 
> file = HdmfFile()
> species = DynamicTable(name='species', description='My species', columns=[col1])
> file.add_container(species)
> ```
> 
> Then in the examples, the `file` argument does not need to be passed because it can be retrieved from the `data` object.
> 
> ```python
> er.add_ref_termset(
> 	container=species,
> 	attribute='Species_Data',
> 	key='Ursus arctos horribilis',
> 	termset=terms
> )
> ```
> 
> This is a more realistic use case, in my opinion. This structure also has the advantage of making the code examples being more standalone / not dependent on earlier objects, except for the definition of `HdmfFile`.
> 
> All of this would help make this tutorial more analogous to its use in PyNWB, so that we could copy and paste relevant parts of this tutorial to create a PyNWB tutorial with minimal modifications, e.g., `HdmfFile` -> `NWBFile`. `add_container` -> `add_acquisition`. Perhaps all of this should be a separate PR though. Up to you. Happy to help with these suggested changes.

I think this is a great idea. I'll make it a separate ticket.

mavaylon1 committed:
name change

mavaylon1 committed:
Update src/hdmf/common/resources.py

Co-authored-by: Ryan Ly <rly@lbl.gov>

mavaylon1 committed:
Update src/hdmf/common/resources.py

Co-authored-by: Ryan Ly <rly@lbl.gov>

mavaylon1 committed:
Update src/hdmf/common/resources.py

Co-authored-by: Ryan Ly <rly@lbl.gov>

mavaylon1 committed:
Update src/hdmf/common/resources.py

Co-authored-by: Ryan Ly <rly@lbl.gov>

mavaylon1 committed:
Update src/hdmf/common/resources.py

Co-authored-by: Ryan Ly <rly@lbl.gov>

mavaylon1 committed:
doc

mavaylon1 committed:
Update src/hdmf/common/resources.py

Co-authored-by: Ryan Ly <rly@lbl.gov>

rly committed:
Merge branch 'dev' into herd_update

mavaylon1 committed:
checks

mavaylon1 committed:
Merge branch 'herd_update' of https://github.com/hdmf-dev/hdmf into herd_update

mavaylon1 committed:
ruff

mavaylon1 committed:
gallery

mavaylon1 committed:
path

mavaylon1 committed:
todo

mavaylon1 committed:
Merge branch 'dev' into herd_update

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

mavaylon1 commented:
@rly A note on your comment of checking that the container is in the file is a good idea. It avoids bad references. I added that change and it would require some major reworking of the test suite. I'd like to make this into a PR that includes the "HDMF_File" as it would go in line with updating the tests. I added the actual code already and a test, these will be flushed out in the next PR. This would also include updates to the doc.

I don't want to manually add the parent in the tests and the documentation then remove them with the HERDFile update.

rly commented on docs/gallery/plot_external_resources.py:
```diff
@@ -166,13 +176,13 @@ def __init__(self, **kwargs):
 
 ###############################################################################
-# Using the add_ref method without the file parameter.
+# Using the add_ref method without the file parametherd.
```
```suggestion
# Using the add_ref method without the file parameter.
```
Be careful with find and replace all!

rly committed:
Update docs/gallery/plot_external_resources.py

rly commented on src/hdmf/common/resources.py:
```diff
@@ -466,64 +476,72 @@ def add_ref_term_set(self, **kwargs):
+            elif isinstance(data_object, (list, tuple, np.ndarray)):
+                data = data_object
+            else:
+                msg = "The data object being used is not supported. Please review the documentation for support types."
```
```suggestion
                msg = "The data object being used is not supported. Please review the documentation for supported types."
```

rly committed:
Update src/hdmf/common/resources.py

rly commented on docs/gallery/plot_external_resources.py:
```diff
@@ -306,14 +316,53 @@ def __init__(self, **kwargs):
+
+# When populating :py:class:`~hdmf.common.resources.HERD`, users may have some terms
+# that are not in the :py:class:`~hdmf.term_set.TermSet`. As a result,
+# :py:func:`~hdmf.common.resources.HERD.add_ref_termset` will return an all the missing
```
```suggestion
# :py:func:`~hdmf.common.resources.HERD.add_ref_termset` will return all of the missing
```

rly committed:
Update docs/gallery/plot_external_resources.py

rly committed:
Update resources.py

rly committed:
Merge branch 'dev' into herd_update
